### ISSUE TITLE
[Question] Handling Negative Samples and Very Fast Convergence with CCT Transformer

### ISSUE THREAD

**Issue by kraster010** [ankandrew/fast-plate-ocr#79]
Hi,
first of all, thank you for this repository.

I successfully retrained the CCT XS transformer model from this repo and I’m very happy with the results.

However, I’m facing one issue: during inference, I sometimes need to process images that do not contain a license plate. In these cases, the transformer still predicts some text, often with very high confidence. This makes it difficult to filter out false positives using a confidence threshold, because lowering the threshold would also remove valid predictions.

Would it make sense, for this type of model, to include negative samples during training?
For example, adding images that do not contain any license plate and associating them with an empty label?
Or would be better to add a second head that do plate/noplate binary classification?

Another observation: training converges very quickly. In about 20 epochs, I reach:

`Validation accuracy > 97%`
`Validation loss < 0.1`

I am using a slightly modified version of the architecture to adapt it to my input resolution (32 × 128).

```
tokenizer:
  blocks:
    - { layer: Conv2D, filters: 32, kernel_size: 3, strides: 1, padding: same, activation: gelu, use_bias: false }
    - { layer: MaxBlurPooling2D, pool_size: 2, filter_size: 3, padding: same }
    - { layer: Conv2D, filters: 48, kernel_size: 3, strides: [ 1, 2 ], padding: same, activation: gelu, use_bias: false }
    - { layer: Conv2D, filters: 64, kernel_size: 3, strides: [ 1, 2 ], padding: same, activation: gelu, use_bias: false }
    - { layer: Conv2D, filters: 80, kernel_size: 3, strides: 1, padding: same, activation: gelu, use_bias: false }

  positional_emb: true
  patch_size: 2
  patch_mlp:
    layer: MLP
    hidden_units: [ 64 ]
    activation: gelu
    dropout_rate: 0.1

transformer_encoder:
  layers: 4
  heads: 1
  projection_dim: 64
  units: [ 64, 64 ]
  activation: gelu
  stochastic_depth: 0.2
  attention_dropout: 0.15
  mlp_dropout: 0.15
  head_mlp_dropout: 0.2
  token_reducer_heads: 4
  normalization: dyt
```
Do you think it would make sense to adjust something (e.g., regularization, training schedule, augmentation, etc.) given this behavior?

For context, the dataset I’m using contains slightly more than 1 million real license plate examples.

Thanks in advance for your feedback 🙂

**Comment by ankandrew** [comment]
Hi! Very good questions and I'm glad it's working well based on initial results.

> Would it make sense, for this type of model, to include negative samples during training?
> For example, adding images that do not contain any license plate and associating them with an empty label?

Yes! For the v2 pre-trained models that are expected to be released soon, I have added negative samples. I have added padding char for all slots for now, to avoid adding extra-head for yes/no `plate`.

> Do you think it would make sense to adjust something (e.g., regularization, training schedule, augmentation, etc.) given this behavior?

Looks good, I think this is a lot of trial and error, although those params look good. I've replaced `gelu` with `silu` in v2 pre-trained models because of similar accuracy and due to export issues w.r.t. some TF versions.

**Comment by kraster010** [comment]
Hi, thanks again for the feedback!

> Looks good, I think this is a lot of trial and error, although those params look good. I've replaced gelu with silu in v2 pre-trained models because of similar accuracy and due to export issues w.r.t. some TF versions.

For now I have not added negative samples to the dataset yet, because after further investigation I realized that part of the issue was actually caused by annotation errors in my dataset. I found and fixed a bug in the auto-annotation procedure I was using to generate the labels. After correcting that and retraining the model, the results improved significantly.

The latest training run produced excellent results, with accuracy comparable to some public commercial OCR libraries, and in several cases even better performance.


Now im starting to work on the detection part that comes before. The detection stage of my pipeline is currently still based on classical computer vision techniques that were state-of-the-art more than a decade ago. I’m interested in experimenting with a more modern CNN-based detector (e.g., YOLO) to see how much the overall system could improve.

However, the target platform is a custom ARMv8 board where I cannot use GPU or any hardware accelerators, so I’m not sure yet whether running a modern detector will be feasible in practice. Still, it seems worth experimenting to evaluate the trade-offs.

I also noticed that you are the maintainer of the FastALPR repository. In the README you mention the model `yolo-v9-t-384-license-plate-end2end` for detection. I was wondering where I could find the code or training setup to retrain that model using my own data.

Is that implementation open source, or is it distributed under a specific license? Also, do you know if there are variants based on newer (or alternatively lighter) YOLO versions that might be suitable for constrained environments like CPU-only ARM systems?

**Comment by ankandrew** [comment]
> ... model yolo-v9-t-384-license-plate-end2end for detection. I was wondering where I could find the code or training setup to retrain that model using my own data.

The `yolo-v9-t-<res>-license-plate-end2end` models can be found (both `.onnx` and `.torch` assets) in https://github.com/ankandrew/open-image-models (specifically [here](https://github.com/ankandrew/open-image-models/releases/tag/assets)). The code to train is the YOLOv9 repo.

I'd try first running a forward pass to see how much time it takes on your ARMv8 with those models. You can also use a model with lower input res (i.e. `256x256`, but if camera is too far away it will have hard time will small plates). If you are running a special LP camera I guess these are not usually very zoomed out, so I guess you can use a model like `yolo-v9-t-256-license-plate-end2end` just fine.

You can also try training nano version of https://github.com/roboflow/rf-detr, better license than `YOLOv5..11`. Honestly the most important thing here is the training data, build a HQ varied data and most of detectors will work well.
